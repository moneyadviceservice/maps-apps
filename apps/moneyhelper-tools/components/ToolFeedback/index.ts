export type { ToolFeedbackProps } from './ToolFeedback';
export { ToolFeedback } from './ToolFeedback';
