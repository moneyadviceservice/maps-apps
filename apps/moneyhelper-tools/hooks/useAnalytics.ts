import { useRouter } from 'next/router';
import { useEffect, useRef } from 'react';

declare global {
  interface Window {
    adobeDataLayer: Array<Record<string, unknown>>;
  }
}

export interface AnalyticsResult {
  addPage: (data: AnalyticsData[]) => void;
  addStepPage: (
    data: AnalyticsPageData,
    currentStep: string | number,
    stepName?: string,
  ) => void;
  addEvent: (data: AnalyticsData) => void;
  analyticsList: React.MutableRefObject<AnalyticsData[]>;
}

export interface AnalyticsData {
  page: {
    lang?: string;
    pageName?: string;
    pageTitle?: string;
    pageType?: string;
    site?: string;
  };
  tool: {
    stepName?: string;
    toolCategory?: string;
    toolName?: string;
    toolStep?: string;
  };
  event?: string;
  errorDetails?: ErrorDetails[];
}

export interface AnalyticsPageData {
  pageName: string;
  pageTitle: string;
  toolName: string;
  stepNames: string[];
}

export interface ErrorDetails {
  fieldType: string | undefined;
  fieldName: string | undefined;
  errorMessage: string | undefined;
}

export const useAnalytics = (): AnalyticsResult => {
  const analyticsList = useRef<AnalyticsData[]>([]);
  const addToLayer = useRef<boolean>(false);
  const router = useRouter();

  useEffect(() => {
    const updateAdobeDataLayer = () => {
      if (
        window.adobeDataLayer &&
        analyticsList.current &&
        !addToLayer.current
      ) {
        analyticsList.current.forEach((page) => {
          addEvent(page);
        });
        addToLayer.current = true;
      }
    };

    updateAdobeDataLayer();
  }, [router.asPath, router.query.language]);

  const addPage = (data: AnalyticsData[]) => {
    data.forEach((d) => {
      analyticsList.current.push({
        event: d.event,
        page: {
          pageName: d.page.pageName ?? '',
          pageTitle: d.page.pageTitle ?? '',
          lang: router.query.language === 'cy' ? 'cy' : 'en',
          site: 'moneyhelper',
          pageType: 'tool page',
        },
        tool: {
          toolName: d.tool.toolName ?? '',
          toolCategory: d.tool.toolCategory ?? '',
          toolStep: d.tool.toolStep ?? '',
          stepName: d.tool.stepName ?? '',
        },
      });
    });
  };

  const addStepPage = (
    analyticsData: AnalyticsPageData,
    currentStep: number | string,
    stepName?: string,
  ) => {
    const step = stepName ?? analyticsData?.stepNames[Number(currentStep) - 1];

    const initialPageData = {
      page: {
        pageName: analyticsData?.pageName,
        pageTitle: analyticsData?.pageTitle,
      },
      tool: {
        toolName: analyticsData?.toolName,
        toolStep: `${currentStep}`,
        stepName: step,
      },
      event: 'pageLoadReact',
    };

    addPage([initialPageData]);
  };

  const addEvent = (data: AnalyticsData) => {
    window.adobeDataLayer.push({
      page: { ...data.page },
      tool: { ...data.tool },
      event: data.event,
    });
  };

  return {
    addPage,
    addStepPage,
    addEvent,
    analyticsList,
  };
};
