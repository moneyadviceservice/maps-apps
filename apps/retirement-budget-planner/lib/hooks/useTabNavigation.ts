import { useRouter } from 'next/router';

export const useTabNavigation = (
  nextTabId: string | undefined,
  onEnableNext: (id: string) => void,
) => {
  const router = useRouter();

  const onNextClick = (
    e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement, MouseEvent>,
  ) => {
    if (nextTabId) {
      e.preventDefault();
      onEnableNext(nextTabId);
      router.push(`/?tab=${nextTabId}`, undefined, { shallow: true });
    }
  };

  return { onNextClick };
};
