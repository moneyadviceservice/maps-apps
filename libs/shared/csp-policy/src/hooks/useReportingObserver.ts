import { useEffect } from 'react';

import { convertCSPReportToString } from '../utils/convertCSPReportToString/convertCSPReportToString';
import { saveDataToBlob } from '../utils/netlify/handleCSPViolationBlob';

/**
 * The useReportingObserver hook is observing for any csp violations and
 * if find any will save them to a netlify blob
 * This hook works only in Chrome, Edge and Safari
 */
export const useReportingObserver = () => {
  const observeViolations = async () => {
    let violations = '';

    if (
      typeof window !== 'undefined' &&
      typeof ReportingObserver !== 'undefined'
    ) {
      const observer = new ReportingObserver(
        async (reports, observer) => {
          reports.forEach((violation) => {
            if (violation.body)
              violations += convertCSPReportToString(violation?.body?.toJSON());
          });
          try {
            if (violations.length > 0) await saveDataToBlob(violations);
          } catch (e) {
            console.warn('Failed to save csp violations to Netlify blobs');
          }
        },
        {
          types: ['csp-violation'],
          buffered: true,
        },
      );

      observer.observe();
    }
  };
  useEffect(() => {
    observeViolations();
  });
};
