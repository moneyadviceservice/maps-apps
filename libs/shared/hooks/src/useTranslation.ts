import { ReactNode } from 'react';
import { TranslationGroup } from './types';
import useLanguage from './useLanguage';

/*
The useTranslation hook that handles all translation inside this app.

Why?

We needed something quick and easy that could handle two languages (only ever two),
and also showed the English right there in the file.

This makes it very easy to find and change text.
It also follows along with the whole tailwind approach of having everything inside the
same file, which is nice.

We did look at other libraries, but not too many of them store the text inside the component.
Also, these libraries are designed for many languages, where we only ever have two.

So, we chose this approach for now. It might seem like a terrible idea, but it should
be easy to replace, and is very easy to understand.

Update Sep 2024: added t() method to load translations from locale json files.

Applications are free to choose between the two methods. Mix them if you like.

Example uses:

(1) Stick to `z()` for shared components where you want to the text to be consistent
for every consumer of the component.

(2) Use the `t()` method when string might be amended by content editors (eg as a sudo CMS).

Usage:

```
// Inline text
const { z } = useTranslation()
return ( <h1> {z({en: "Hello World", cy: "Hello World"})} </h1> )

// Object interpolation
const { z } = useTranslation()
return ( <h1> {z({en: "Hello {thing}", cy: "Helo {thing}"}, { thing: "world"})} </h1> )

// JSX
const { z } = useTranslation()
return ( <h1> {z({en: <div>hello</div>, cy: <div>helo</div>})} </h1> )

// Locale json file
const { t } = useTranslation()
return ( <h1> {t("helloWorld")} </h1> )

// Locale json with interpolation
const { t } = useTranslation()
return ( <h1> {t("helloWorld", { thing: "world"})} </h1> )

```

If there is no cy translation then just leave it out, and some console errors will be shown.
You can then hand these over for translation.
*/

type Data = Record<string, string>;
const fallbackText = '<No translation available>';

type FlattenJsonOptions = {
  data: Record<string, unknown>;
  parentKey?: string;
  result?: Record<string, string>;
};

const flattenJson = ({
  data,
  parentKey = '',
  result = {},
}: FlattenJsonOptions): Record<string, string> => {
  for (const key in data) {
    if (Object.prototype.hasOwnProperty.call(data, key)) {
      const newKey = parentKey ? `${parentKey}.${key}` : key;

      // If the value is an object, recursively flatten it
      const isObject =
        data[key] && typeof data[key] === 'object' && !Array.isArray(data[key]);
      if (isObject) {
        flattenJson({
          data: data[key] as Record<string, unknown>,
          parentKey: newKey,
          result,
        });
      } else {
        result[newKey] = String(data[key]);
      }
    }
  }
  return result;
};

export const useTranslation = (localeOverirde?: string) => {
  let locale = useLanguage();
  if (localeOverirde) {
    locale = localeOverirde;
  }

  if (!(locale === 'en' || locale === 'cy')) {
    throw new Error(`locale [${locale}] does not exist`);
  }

  const interpolate = ((node: ReactNode, data?: Data) => {
    if (typeof node !== 'string') {
      return node;
    }
    if (!data) {
      return node;
    }
    return Object.keys(data).reduce((previousValue, currentValue) => {
      if (typeof previousValue !== 'string') return node;
      return previousValue.replaceAll(`{${currentValue}}`, data[currentValue]);
    }, node);
  }) as <T extends ReactNode>(
    node: T,
    data?: Data,
  ) => T extends string ? string : T;

  const findFirstAvailableTranslation = (translation: TranslationGroup) => {
    const firstKey = Object.keys(translation)[0];
    return translation[firstKey as keyof TranslationGroup];
  };

  const z = ((
    translation: TranslationGroup,
    data?: Data,
    fallback = fallbackText,
  ) => {
    const value = translation[locale as keyof TranslationGroup];
    if (value) {
      return interpolate(value, data);
    }

    console.warn(
      `no translation available in [${locale}] for [${findFirstAvailableTranslation(
        translation,
      )}]`,
    );
    return fallback;
  }) as <T extends TranslationGroup>(
    translation: T,
    data?: Data,
    fallback?: string,
  ) => T[keyof T] extends string ? string : T[keyof T];

  const t = (key: string, data?: Data, fallback = fallbackText) => {
    try {
      // import the translations from the applications /public/locales folder
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const translations = require(`/public/locales/${locale}.json`);
      const value = flattenJson({ data: translations })[key];

      if (value) {
        return interpolate(value, data);
      }

      return fallback;
    } catch (e) {
      console.warn(
        `no translation available in /public/locales/${locale}.json for [${key}]`,
      );
      return fallback;
    }
  };

  return {
    z,
    t,
    locale,
  } as const;
};

export default useTranslation;
