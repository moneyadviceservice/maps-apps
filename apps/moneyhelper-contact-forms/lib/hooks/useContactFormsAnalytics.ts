import { useEffect, useMemo } from 'react';

import {
  AnalyticsData,
  EventErrorDetails,
  useAnalytics,
} from '@maps-react/hooks/useAnalytics';
import useTranslation from '@maps-react/hooks/useTranslation';
import { formatAnalyticsObject } from '@maps-react/utils/formatAnalyticsObject';
import { AnalyticsToolData } from '@maps-react/utils/formatAnalyticsObject/formatAnalyticsObject';

import { StepName } from '../constants';
import { Entry, FormError } from '../types';

export type useContactFormsAnalyticsProps = {
  step: string;
  entry?: Entry;
  errors?: FormError;
  referenceNumber?: string;
  url: string | undefined;
};

/**
 * Custom hook to handle analytics for contact forms.
 * It captures page load events, errors, and formats the analytics data.
 * @param {string} - The current step in the contact form.
 * @param {Entry}  - The entry data for the contact form.
 * @param {FormError[]} - An array of errors encountered in the form.
 * @param {string} - The reference number for the contact form submission (confirmation page only).
 * @param {string} - The URL of the current page
 * @return {void}
 */
export function useContactFormsAnalytics({
  step,
  entry,
  errors,
  referenceNumber,
  url,
}: useContactFormsAnalyticsProps) {
  const { t: tEn, z: zEn } = useTranslation('en');
  const { t: tCy } = useTranslation('cy');
  const { addEvent, addPage } = useAnalytics();

  const pageTitle =
    step.replace(/-/g, ' ').charAt(0).toUpperCase() +
    step.replace(/-/g, ' ').slice(1); // make the pageTitle look nice
  const enquiryForm = 'enquiry-form';
  const stepIndex = entry ? entry.stepIndex + 1 : 0;
  const flow = entry?.data?.flow ?? '';
  const pageName = flow ? `${enquiryForm}-${flow}` : `${enquiryForm}`;

  const anylticsToolData: AnalyticsToolData = useMemo(
    () => ({
      tool: tEn('site.title'),
      toolCy: tCy('site.title'),
      toolStep: stepIndex.toString(),
      stepData: {
        pageName: step,
        pageTitle,
        stepName: step,
      },
      pageToolName: pageName,
      categoryLevels: [enquiryForm],
      url,
    }),
    [tEn, tCy, stepIndex, step, pageTitle, pageName, enquiryForm, url],
  );

  const analyticsData: AnalyticsData = useMemo(
    () => ({
      ...formatAnalyticsObject(zEn, anylticsToolData),
      event: 'pageLoadReact',
      tool: {
        ...formatAnalyticsObject(zEn, anylticsToolData).tool,
        toolCategory: flow,
        ...(referenceNumber &&
          step === StepName.CONFIRMATION && {
            outcome: {
              ID: referenceNumber,
            },
          }),
      },
      ...(entry?.data.year &&
        step === StepName.CONTACT_DETAILS && {
          demo: {
            bYear: parseInt(entry.data.year, 10),
          },
        }),
    }),
    [zEn, anylticsToolData, flow, referenceNumber, entry, step],
  );

  // Create the error event - if the step is 'error', we log the error message directly (passed in on error.tsx),
  const errorDetails: EventErrorDetails[] = useMemo(() => {
    if (!errors || Object.keys(errors).length === 0) return [];

    if (step === StepName.ERROR) {
      const errorMsg = errors['error-page']?.[0] ?? '';
      return [
        {
          reactCompType: tEn(`components.${step}.title`),
          reactCompName: 'error-page',
          errorMessage: errorMsg,
        },
      ];
    }

    // For all other steps, flatten all error keys for each field
    return Object.entries(errors).flatMap(([field, errorKeys]) =>
      errorKeys.map((errorKey) => ({
        reactCompType: tEn(`components.${step}.title`),
        reactCompName: field,
        errorMessage: tEn(`components.${step}.form.${errorKey}.error`),
      })),
    );
  }, [errors, tEn, step]);

  useEffect(() => {
    addEvent(analyticsData);
    if (errorDetails.length > 0) {
      addEvent({
        event: 'errorMessage',
        eventInfo: {
          errorDetails,
        },
      });
    }
  }, [addPage, addEvent, analyticsData, errorDetails, step, entry]);
}
