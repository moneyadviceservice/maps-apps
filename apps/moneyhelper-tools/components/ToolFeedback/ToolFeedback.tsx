import { twMerge } from 'tailwind-merge';

import { useTranslation } from '@maps-react/hooks/useTranslation';

import styles from './styles/widget.module.scss';

export type ToolFeedbackProps = {
  className?: string;
  id: string;
};

export const ToolFeedback = ({ className, id }: ToolFeedbackProps) => {
  const { locale } = useTranslation();

  return (
    <div
      className={twMerge(
        `${styles.toolFeedbackWidget} mt-8`,
        locale === 'cy' ? styles.welshBackButton : '',
        className,
      )}
      id={id}
    >
      {/* Injected widget will appear here */}
    </div>
  );
};
