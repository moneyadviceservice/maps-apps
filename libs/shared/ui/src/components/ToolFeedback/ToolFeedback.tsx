import { useEffect } from 'react';

import { twMerge } from 'tailwind-merge';

import { useTranslation } from '@maps-react/hooks/useTranslation';

import styles from './styles/widget.module.scss';
declare global {
  interface Window {
    informizelyApi?: {
      showScript: (surveyId: string) => void;
    };
  }
}
export type ToolFeedbackProps = {
  className?: string;
  surveyIds?: {
    production: {
      en: string;
      cy: string;
    };
    development: {
      en: string;
      cy: string;
    };
  };
  overrideIzCr?: boolean;
};

const DEFAULT_SURVEY_IDS = {
  production: {
    en: 'informizely-embed-fgddnriui',
    cy: 'informizely-embed-ugeryrudd',
  },
  development: {
    en: 'informizely-embed-fjguluwfj',
    cy: 'informizely-embed-zjiieufr',
  },
};

export const ToolFeedback = ({
  className,
  surveyIds = DEFAULT_SURVEY_IDS,
  overrideIzCr,
}: ToolFeedbackProps) => {
  const { locale } = useTranslation();

  const currentIds =
    process.env.NEXT_PUBLIC_ENVIRONMENT === 'production'
      ? surveyIds.production[locale]
      : surveyIds.development[locale];

  useEffect(() => {
    const timer = setTimeout(() => {
      if (typeof window !== 'undefined' && window.informizelyApi) {
        const apiSurveyId = currentIds.replace('informizely-embed-', '');
        window.informizelyApi.showScript(apiSurveyId);
      }
    }, 500);
    return () => clearTimeout(timer);
  }, [currentIds]);

  return (
    <div
      className={twMerge(
        `${styles.toolFeedbackWidget} mt-8`,
        locale === 'cy' ? styles.welshBackButton : '',
        overrideIzCr ? styles.overrideIzCr : '',
        className,
      )}
      id={currentIds}
    />
  );
};
