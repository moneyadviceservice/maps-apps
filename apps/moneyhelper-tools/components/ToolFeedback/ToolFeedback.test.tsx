import { render } from '@testing-library/react';

import { useTranslation } from '@maps-react/hooks/useTranslation';

import { ToolFeedback } from './ToolFeedback';

jest.mock('@maps-react/hooks/useTranslation');

describe('ToolFeedback component', () => {
  it('renders correctly (en)', () => {
    (useTranslation as jest.Mock).mockReturnValue({
      locale: 'en',
    });

    const { container } = render(<ToolFeedback id="test-id" />);
    expect(container.firstChild).toMatchSnapshot();

    jest.clearAllMocks();
  });

  it('renders correctly (cy)', () => {
    (useTranslation as jest.Mock).mockReturnValue({
      locale: 'cy',
    });

    const { container } = render(<ToolFeedback id="test-id" />);
    expect(container.firstChild).toMatchSnapshot();

    jest.clearAllMocks();
  });
});
