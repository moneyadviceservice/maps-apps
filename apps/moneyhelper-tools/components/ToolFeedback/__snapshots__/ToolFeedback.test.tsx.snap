// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolFeedback component renders correctly (cy) 1`] = `
<div
  class="toolFeedbackWidget mt-8 welshBackButton"
  id="test-id"
/>
`;

exports[`ToolFeedback component renders correctly (en) 1`] = `
<div
  class="toolFeedbackWidget mt-8"
  id="test-id"
/>
`;
